### PR TITLE
Users/padepu/azure web app acr support

### PR DESCRIPTION
**Pull request by pavanadepu2**
Added the support for Azure Container Registry..

### CODE REVIEW

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 1]
```diff
@@ -13,28 +13,35 @@ var restObj = new restClient.RestCallbackClient(httpObj);
 
 var defaultAuthUrl = 'https://login.windows.net/';
 var azureApiVersion = 'api-version=2016-08-01';
+var azureContainerRegistryApiVersion = "api-version=2017-03-01";
```
Comment (kmkumaran) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: @Ajay-MS is there a support for Azure stack?

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 2]
```diff
@@ -182,7 +189,7 @@ function getAuthorizationToken(endPoint): Q.Promise<string> {
     return deferred.promise;
 }
 
-async function getAzureRMWebAppID(endpoint, webAppName: string, url: string, headers) {
+async function getAzureRMResourceID(endpoint, resourceName: string, url: string, headers) {
```
Comment (kmkumaran) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: @arjgupta - Is there anything to react here for the proxy changes that you are doing in parallel?

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 3]
```diff
@@ -112,17 +129,99 @@
...
+            "defaultValue": "",
+            "required": true,
+            "properties": {
+                "EditableOptions": "True"
+            },
+            "helpMarkDown": "A globally unique top-level domain name for your specific registry.<br/> Note: Fully qualified image name will be of the format: '<b>`<registry`></b>/`<repository`>:`<tag`>'. For example, '<b>myregistry.azurecr.io</b>/nginx:latest'.",
+            "visibleRule": "ImageSource = AzureContainerRegistry"
+        },
+        {
+            "name": "AzureContainerRegistryLoginServer",
```
[the remark this answers is not in the document]
Comment (pavanadepu2) [reply Tasks/AzureRmWebAppDeployment/task.json]: This is the LoginServerUrl. SPN is just creds, we need registry server url to connect to a registry.

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 4]
```diff
@@ -112,17 +129,99 @@
...
-            "helpMarkDown": "Select the option to choose the source of the image for your web-app.",
-            "visibleRule": "WebAppKind = applinux"
+            "defaultValue": "public",
+            "helpMarkDown": "Select the Docker repository access.",
+            "visibleRule": "ImageSource = invalidImage"
+        },
+        {
+            "name": "RegistryConnectedServiceName",
+            "type": "connectedService:dockerregistry",
+            "label": "Registry Connection",
```
[the remark this answers is not in the document]
Comment (pavanadepu2) [reply Tasks/AzureRmWebAppDeployment/task.json]: This is applicable only for dockerhub and private. The "Registry Connection" makes sense as it is inline with Azure Connection.

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 5]
```diff
@@ -79,6 +79,23 @@
             }
         },
         {
+            "name": "ImageSource",
```
[the remark this answers is not in the document]
Comment (pavanadepu2) [reply Tasks/AzureRmWebAppDeployment/task.json]: Seems an issue with auto merge, corrected the order.

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 6]
```diff
@@ -136,10 +235,10 @@
         {
             "name": "DockerRepository",
             "type": "string",
-            "label": "Repository",
+            "label": "Image",
```
[the remark this answers is not in the document]
Comment (pavanadepu2) [reply Tasks/AzureRmWebAppDeployment/task.json]: Reverting it to Repository. Will take this change while supporting all registries.

File: Tasks/Common/azurerest-common/azurerestutility.ts (hunk 1 above)
Comment (Ajay-MS) [reply Tasks/Common/azurerest-common/azurerestutility.ts]: Endpoint object initialization takes care of National Cloud/Azure Stack  URLs. Hence changes are compatible for the azure stack and national cloud. We might have to change API version for azure stack later as currently, ACR support is not there with Azure Stack and not sure about API version.   @pavanadepu2  Can you please verify changes for US government cloud once.

File: Tasks/Common/azurerest-common/azurerestutility.ts (hunk 2 above)
Comment (arjgupta) [reply Tasks/Common/azurerest-common/azurerestutility.ts]: changes to the rest utility will need to be incorporated in my PR. I will do it once these changes are merged. I don't see anything that needs to be change as far as this PR is concerned.

**Pull request merged**